src/lua/uri.c

```diff
@@ -0,0 +1,390 @@
+	 * format.
+	 */
+	for (int i = 0; i < uri_count && rc == 0; i++) {
+		lua_rawgeti(L, idx, i + 1);
+		switch (lua_type(L, -1)) {
+		case LUA_TSTRING:
+		case LUA_TNUMBER:
+			if (uri_set_add(uri_set, lua_tostring(L, -1)) == NULL) {
+				errmsg = "Incorrect URI: expected host:service or "
+					 "/unix.socket";
+				diag_set(IllegalParams, errmsg);
+				rc = -1;
+			}
+			break;
+		case LUA_TTABLE:
+			uri = uri_set_add(uri_set, NULL);
+			assert(uri != NULL);
+			rc = uri_create_from_lua_table(uri, L, lua_gettop(L));
+			assert(rc == 0 || diag_get() != NULL);
+			break;
+		default:
+			errmsg = "Incorrect type for URI in table: "
+				 "should be string, number or table";
+			diag_set(IllegalParams, errmsg);
+			rc = -1;
+		}
+		lua_pop(L, 1);
+	}
+	if (rc != 0)
+		return rc;
+
+	/*
+	 * Here we are only in case when uri_count greater than one,
+	 * so is shouldn't be "uri" and "params" field here.
+	 */
+	if (is_field_present(L, "uri", idx)) {
+		diag_set(IllegalParams, "Incorrect URI count in table");
+		return -1;
+	}
+
+	if (is_field_present(L, "params", idx)) {
+		diag_set(IllegalParams, "URI query parameters are "
+			 "not allowed for multiple URIs");
+		return -1;
+	}
+
+	for (int i = 0; i < uri_set->uri_count && rc == 0; i++) {
+		uri = &uri_set->uris[i];
+		rc = uri_add_params_from_lua_table(uri, L, "default_params",
+						   idx, false);
+		assert(rc == 0 || diag_get() != NULL);
+	}
+	return rc;
+}
+
+/**
+ * Create @uri fromthe value at the given valid @a idx
+ */
+static int
+luaT_uri_create(struct lua_State *L, struct uri *uri, int idx)
+{
+	const char *errmsg;
+	int rc = 0;
+	uri_create(uri, NULL);
+	if (lua_isstring(L, idx)) {
+		rc = uri_create(uri, lua_tostring(L, idx));
+		if (rc != 0) {
+			errmsg = "Incorrect URI: expected host:service or "
+				 "/unix.socket";
+			diag_set(IllegalParams, errmsg);
+		}
+	} else if (lua_istable(L, idx)) {
+		rc = uri_create_from_lua_table(uri, L, idx);
+	} else if (!lua_isnil(L, idx)) {
+		errmsg = "Incorrect type for URI: "
+			 "should be string, number or table";
+		diag_set(IllegalParams, errmsg);
+		rc = -1;
+	}
+	if (rc == -1) {
+		assert(diag_get() != NULL);
+		uri_destroy(uri);
+	}
+	return rc;
+}
+
+/**
+ * Create @uri_set from the value at the given valid @a idx.
+ */
+static int
+luaT_uri_set_create(struct lua_State *L, struct uri_set *uri_set, int idx)
+{
+	const char *errmsg;
+	int rc = 0;
+	uri_set_create(uri_set, NULL);
+	if (lua_isstring(L, idx)) {
+		rc = uri_set_create(uri_set, lua_tostring(L, idx));
+		if (rc != 0) {
+			errmsg = "Incorrect URI: expected host:service or "
+				 "/unix.socket";
+			diag_set(IllegalParams, errmsg);
+		}
+	} else if (lua_istable(L, idx)) {
+		rc = uri_set_create_from_lua_table(uri_set, L, idx);
+	} else if (!lua_isnil(L, idx)) {
+		errmsg = "Incorrect type for URI: "
+			 "should be string, number or table";
+		diag_set(IllegalParams, errmsg);
+		rc = -1;
+	}
+	if (rc == -1) {
+		assert(diag_get() != NULL);
+		uri_set_destroy(uri_set);
```

Comment:
Why? It should have been destroyed by the callee on error, no? Same question for `uri_destroy` in `luaT_uri_create`.

Reply:
Yes you right.